### PR TITLE
make internal solution for cf616d Longest k-Good Segment

### PR DESCRIPTION
**Pull request by TruaShamu**

### CODE REVIEW

File: solutions/cf-616D.mdx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+<JavaSection>
+
+## Solution
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+## Java Implementation
+
+```java
+public class kgood {
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:21]:
```suggestion
public class KGood {
```

File: solutions/cf-616D.mdx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                maxLen = curLen;
+                maxL = left;
+                maxR = i;
+            }
+        }
+
+        System.out.println((maxL + 1) + " " + (maxR + 1));
+
+    }
+
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:72]: lines 25, 39, 40, 68, 70, and 72 are unnecessary whitespace

File: solutions/cf-616D.mdx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+}
+
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:74]: oh yeah also line 74

File: solutions/cf-616D.mdx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	CodeSnip{Kattio}
+
+    public static int MAX = (int) 1e6;
+
+    public static void main(String[] args) throws IOException {
+        Kattio io = new Kattio();
+        int N = io.nextInt();
+        int K = io.nextInt();
+
+        int[] A = new int[MAX];
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:31]: it'd be cool if you could
1. replace `MAX` with `N` (i mean arrays in java don't have to constant size)
2. rename this to `arr` (just a tiny bit more explicit)

File: solutions/cf-616D.mdx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        int[] freq = new int[MAX]; // Frequency of each number in the window.
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:32]: 1. again, replace `MAX` with `N`
2. i think the style is 2 lines after a line of code for a comment (like `int x = 1;  // this is a comment`)

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
yeah it's p cool just some minor things

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
use tabs

**Comment by TruaShamu** [comment]
Ok, it should be fixed.

I'm sorry, I don't really understand what you mean about the code comments though... Should I put the comments on different lines, or?

**Comment by SansPapyrus683** [comment]
add an extra space- rn your code comments are like so:
```java
int x = 1; // this is a comment
```
i'd suggest adding a space between the code and the comment like this:
```java
int x = 1;  // this is a better comment
```

**Comment by TruaShamu** [comment]
I reformatted the comments :)

File: solutions/cf-616D.mdx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+public class KGood {
+	CodeSnip{Kattio}
+
+	public static int MAX = (int) 1e6;
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio();
+		int N = io.nextInt();
+		int K = io.nextInt();
+		int[] arr = new int[N];
+		int[] freq = new int[MAX];  // Frequency of each number in the window.
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:30]: uh why not make `freq` also just size `N` as well?

Comment (TruaShamu) [reply solutions/cf-616D.mdx:30]: In the [problem,](https://codeforces.com/problemset/problem/616/D) the values of the array `arr` aren't bounded between 1 and N, they're bounded between (0 ≤ arr[i] ≤ 10^6)

The frequency array `freq` contains the frequency of each number `arr[i]` , therefore, I think that the length of `freq[]` should be the maximum value in `arr[i]`.
And because I was lazy and new that the max value of `arr[i]` is 10^6, I set the length of `freq` to 10^6.

If you want, I could add a for loop to find the maximum value of arr[i] and then set that as the length of freq, I guess?

(Sorry if I'm being confusing, and feel free to correct me if I'm wrong )

Comment (SansPapyrus683) [reply solutions/cf-616D.mdx:30]: oof sorry i didn't read it clear enough
> If you want, I could add a for loop to find the maximum value of arr[i] and then set that as the length of freq, I guess?

this is what i'd do tbh (a oneliner would be `max = Arrays.stream(arr).max().getAsInt()`) but i'm not sure what the other people want

Comment (TruaShamu) [reply solutions/cf-616D.mdx:29]: Ok, I fixed it :)

File: solutions/cf-616D.mdx [hunk 7]
```diff
@@ -18,24 +18,30 @@ author: Sofia Yang
...
 		int maxLen = 0;
 		int maxL = 0;
 		int maxR = 0;
+		int[] arr = new int[N];
+		for (int i = 0; i < N; i++) {
+					arr[i] = io.nextInt();
+			}
+			
+		int max = Arrays.stream(A).max().getAsInt() + 1;
+		int[] freq = new int[max];  // Frequency of each number in the window.
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:41]: replace these two lines w/
```java
		int max = Arrays.stream(A).max().getAsInt();
		int[] freq = new int[max + 1];  // Frequency of each number in the window.
```

File: solutions/cf-616D.mdx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+```java
+import java.util.*;
+public class KGood {
+	CodeSnip{Kattio}
+
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio();
+		int N = io.nextInt();
+		int K = io.nextInt();
+		int[] freq = new int[MAX];  // Frequency of each number in the window.
```
Comment (dongliu0426) [inline solutions/cf-616D.mdx:29]: does this compile? what is `MAX`

Comment (SansPapyrus683) [reply solutions/cf-616D.mdx:29]: it's a constant, i told her to change it to something that's dynamically calculated, dw about it

Comment (TruaShamu) [reply solutions/cf-616D.mdx]: I forgot to delete it from earlier, it's gone now.

File: solutions/cf-616D.mdx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+			// Update the window length.
+			int curLen = i - left + 1;
+			if (curLen > maxLen) {
+				maxLen = curLen;
+				maxL = left;
+				maxR = i;
+			}
+		}
+		System.out.println((maxL + 1) + " " + (maxR + 1));
```
Comment (SansPapyrus683) [inline solutions/cf-616D.mdx:70]:
```suggestion
		System.out.printf("%d %d%n", maxL + 1, maxR + 1);
```
honestly up to you though

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/cf-616D.mdx [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+			// Update the window length.
+			int curLen = i - left + 1;
+			if (curLen > maxLen) {
+				maxLen = curLen;
+				maxL = left;
+				maxR = i;
+			}
+		}
+		System.out.printf("%d %d%n", maxL + 1, maxR + 1);
```
Comment (dongliu0426) [inline solutions/cf-616D.mdx:70]: use kattio for this?

Comment (SansPapyrus683) [reply solutions/cf-616D.mdx:70]: is there really a need for that

Comment (TruaShamu) [reply solutions/cf-616D.mdx]: there, is everyone satisfied?

File: solutions/cf-616D.mdx [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+<JavaSection>
+
+## Solution
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+## Java Implementation
+
+```java
+import java.util.*;
```
Comment (dongliu0426) [inline solutions/cf-616D.mdx:21]:
```suggestion
import java.io.*;
import java.util.*;
```

File: solutions/cf-616D.mdx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		int left = 0;
+		int maxLen = 0;
+		int maxL = 0;
+		int maxR = 0;
+		int[] arr = new int[N];
+		for (int i = 0; i < N; i++) {
+			arr[i] = io.nextInt();
+		}
+			
+		int max = Arrays.stream(A).max().getAsInt();
```
Comment (dongliu0426) [inline solutions/cf-616D.mdx:39]:
```suggestion
		int max = Arrays.stream(arr).max().getAsInt();
```
???

File: solutions/cf-616D.mdx [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+			// Update the window length.
+			int curLen = i - left + 1;
+			if (curLen > maxLen) {
+				maxLen = curLen;
+				maxL = left;
+				maxR = i;
+			}
+		}
+		io.printf("%d %d%n", maxL + 1, maxR + 1);
```
Comment (dongliu0426) [inline solutions/cf-616D.mdx:70]:
```suggestion
		io.printf("%d %d%n", maxL + 1, maxR + 1);
		io.close();
```

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/cf-616D.mdx (hunk 11 above)
Comment (dongliu0426) [reply solutions/cf-616D.mdx:21]: doesn't compile w/o this

**Review by dongliu0426 — APPROVED** [review]

**Comment by darren-yao** [comment]
Approved and merged.
